File: lib/formatters.rb
# frozen_string_literal: true

require "time"
require_relative "alttext_api"

module Formatters
  module_function

  def format_image(image)
    lines = []
    lines << "Asset ID: #{image["asset_id"]}"
    lines << "Alt text: #{image["alt_text"] || "(none)"}"
    lines << "URL: #{image["url"]}" if image["url"]

    alt_texts = image["alt_texts"] || {}
    if alt_texts.any?
      lines << "Languages:"
      alt_texts.each { |lang, text| lines << "  #{lang}: #{text}" }
    end

    lines << "Tags: #{image["tags"].join(", ")}" if image["tags"]&.any?
    lines << "Metadata: #{JSON.generate(image["metadata"])}" if image["metadata"]&.any?

    if image["created_at"].is_a?(Numeric)
      lines << "Created: #{Time.at(image["created_at"]).utc.iso8601}"
    end

    errors = image["errors"] || {}
    if errors.any?
      lines << "Errors: #{errors.values.flatten.join(", ")}"
    end

    lines.join("\n")
  end

  def format_image_list(result)
    pagination = result[:pagination]
    images = result[:images]

    lines = ["Found #{pagination[:total_count]} images (page #{pagination[:current_page]} of #{pagination[:total_pages]})", ""]

    if images.empty?
      lines << "No images found."
    else
      images.each do |image|
        lines << "- #{image["asset_id"]}: #{image["alt_text"] || "(no alt text)"}"
      end
    end

    lines.join("\n")
  end

  def format_account(account)
    credits_remaining = (account["usage_limit"] || 0) - (account["usage"] || 0)

    lines = [
      "Account: #{account["name"]}",
      "Credits: #{credits_remaining} remaining (#{account["usage"]} used of #{account["usage_limit"]})",
      "Default language: #{account["default_lang"]}",
    ]
    lines << "Custom prompt: #{account["gpt_prompt"]}" if account["gpt_prompt"]
    lines << "Max chars: #{account["max_chars"]}" if account["max_chars"]
    lines << "Webhook: #{account["webhook_url"]}" if account["webhook_url"]
    lines << "Notification email: #{account["notification_email"]}" if account["notification_email"]
    lines << "Whitelabel: #{account["whitelabel"]}"
    lines << "No quotes: #{account["no_quotes"]}"
    lines << "Ending period: true" if account["ending_period"]

    lines.join("\n")
  end

  def format_scrape_result(result, url)
    scraped = result["scraped_images"] || []

    lines = [
      "Scraped #{url}",
      "Images found: #{scraped.length}",
      "Images queued for processing: #{result["total_processed"] || 0}",
      "",
    ]

    if scraped.any?
      lines << "Discovered images:"
      scraped.each do |img|
        status = img["skip_reason"] ? "skipped: #{img["skip_reason"]}" : "queued"
        lines << "  - #{img["src"] || "(no src)"} [#{status}]"
      end
    end

    errors = result["errors"] || {}
    if errors.any?
      lines << "\nErrors: #{errors.values.flatten.join(", ")}"
    end

    if (result["total_processed"] || 0).positive?
      lines << "\nNote: Images are being processed asynchronously. Use list_images or get_image to check results."
    end

    lines.join("\n")
  end

  def error_response(error)
    if error.is_a?(AltTextApiError)
      parts = ["Error (#{error.status}): #{error.message}"]
      parts << "Code: #{error.error_code}" if error.error_code
      MCP::Tool::Response.new([{ type: "text", text: parts.join("\n") }], error: true)
    else
      MCP::Tool::Response.new([{ type: "text", text: "Error: #{error.message}" }], error: true)
    end
  end
end

File: src/formatters.ts
import type { AccountRecord, ImageListResult, ImageRecord, ScrapeResult } from "./types.js";

/** Format a single image record for display. */
export function formatImage(image: ImageRecord): string {
  const lines: string[] = [];
  lines.push(`Asset ID: ${image.asset_id}`);
  lines.push(`Alt text: ${image.alt_text ?? "(none)"}`);

  if (image.url) {
    lines.push(`URL: ${image.url}`);
  }

  if (image.alt_texts && Object.keys(image.alt_texts).length > 0) {
    lines.push("Languages:");
    for (const [lang, text] of Object.entries(image.alt_texts)) {
      lines.push(`  ${lang}: ${text}`);
    }
  }

  if (image.tags && image.tags.length > 0) {
    lines.push(`Tags: ${image.tags.join(", ")}`);
  }

  if (image.metadata && Object.keys(image.metadata).length > 0) {
    lines.push(`Metadata: ${JSON.stringify(image.metadata)}`);
  }

  if (typeof image.created_at === "number") {
    lines.push(`Created: ${new Date(image.created_at * 1000).toISOString()}`);
  }

  if (image.error_code) {
    lines.push(`Error code: ${image.error_code}`);
  }

  if (image.errors && Object.keys(image.errors).length > 0) {
    lines.push(`Errors: ${Object.values(image.errors).flat().join(", ")}`);
  }

  return lines.join("\n");
}

/** Format a paginated image list for display. */
export function formatImageList(result: ImageListResult): string {
  const { pagination, images } = result;
  const lines: string[] = [
    `Found ${String(pagination.totalCount)} images (page ${String(pagination.currentPage)} of ${String(pagination.totalPages)})`,
    "",
  ];

  if (images.length === 0) {
    lines.push("No images found.");
  } else {
    for (const image of images) {
      lines.push(`- ${image.asset_id}: ${image.alt_text ?? "(no alt text)"}`);
    }
  }

  return lines.join("\n");
}

/** Format account information for display. */
export function formatAccount(account: AccountRecord): string {
  const creditsRemaining = account.usage_limit - account.usage;

  const lines: string[] = [
    `Account: ${account.name}`,
    `Credits: ${String(creditsRemaining)} remaining (${String(account.usage)} used of ${String(account.usage_limit)})`,
    `Default language: ${account.default_lang}`,
  ];

  if (account.gpt_prompt) lines.push(`Custom prompt: ${account.gpt_prompt}`);
  if (account.max_chars) lines.push(`Max chars: ${String(account.max_chars)}`);
  if (account.webhook_url) lines.push(`Webhook: ${account.webhook_url}`);
  if (account.notification_email) lines.push(`Notification email: ${account.notification_email}`);
  lines.push(`Whitelabel: ${String(account.whitelabel)}`);
  lines.push(`No quotes: ${String(account.no_quotes)}`);
  if (account.ending_period) lines.push("Ending period: true");

  return lines.join("\n");
}

/** Format page scrape results for display. */
export function formatScrapeResult(result: ScrapeResult, url: string): string {
  const scraped = result.scraped_images;

  const lines: string[] = [
    `Scraped ${url}`,
    `Images found: ${String(scraped.length)}`,
    `Images queued for processing: ${String(result.total_processed)}`,
    "",
  ];

  if (scraped.length > 0) {
    lines.push("Discovered images:");
    for (const img of scraped) {
      const status = img.skip_reason ? `skipped: ${img.skip_reason}` : "queued";
      const dims =
        img.width !== undefined && img.height !== undefined
          ? ` (${String(img.width)}x${String(img.height)})`
          : "";
      lines.push(`  - ${img.src ?? "(no src)"}${dims} [${status}]`);
    }
  }

  if (result.errors && Object.keys(result.errors).length > 0) {
    lines.push(`\nErrors: ${Object.values(result.errors).flat().join(", ")}`);
  }

  if (result.total_processed > 0) {
    lines.push(
      "\nNote: Images are being processed asynchronously. Use list_images or get_image to check results.",
    );
  }

  return lines.join("\n");
}
